### ISSUE TITLE
Consider NFC normalization of idents?

### ISSUE THREAD

**Issue by workingjubilee** [rust-lang/rustfmt#6058]
Raised by @Jules-Bertholet in https://github.com/rust-lang/rust/issues/120697

It could be desirable to NFC-normalize identifiers so as to remove these kinds of inconsistencies.

However, it seems plausible that this may cause issues if the identifiers are not local, e.g. they come from another crate that does not run rustfmt.

It may make sense, however, if rustc itself linted against non-NFC-normalized idents?

**Comment by Jules-Bertholet** [comment]
> this may cause issues if the identifiers are not local, e.g. they come from another crate that does not run rustfmt

I don't see how?

**Comment by workingjubilee** [comment]
@Jules-Bertholet Wait, maybe my brain is just slow (I'm overdue for lunch...), does rustc fully NFC-normalize idents internally when resolving them?

**Comment by Jules-Bertholet** [comment]
> @Jules-Bertholet Wait, maybe my brain is just slow (I'm overdue for lunch...), does rustc fully NFC-normalize idents internally when resolving them?

Yes.

**Comment by workingjubilee** [comment]
Oh, no issue then!

**Comment by ytmimi** [comment]
@Jules-Bertholet @workingjubilee Do either of you know if there are crates that already implement the normalization?

**Comment by Jules-Bertholet** [comment]
Rustc uses [`unicode-normalization`](https://crates.io/crates/unicode-normalization)

**Comment by ytmimi** [comment]
@Jules-Bertholet Ahhh nice! Very good to know that something like this already exists. Thanks for the link.

> However, it seems plausible that this may cause issues if the identifiers are not local, e.g. they come from another crate that does not run rustfmt.

Just want to make sure I understand correctly. Neither of you think this would be an issue?

**Comment by Jules-Bertholet** [comment]
> Just want to make sure I understand correctly. Neither of you think this would be an issue?

Correct, AIUI. Even if a dependency uses non-NFC idents internally, they can be referred to by dependents with their NFC forms.

**Comment by workingjubilee** [comment]
> Just want to make sure I understand correctly. Neither of you think this would be an issue?

Yes. I was a bit slow on the uptake and hadn't remembered yet that we NFC idents (because of course we should). Applying NFC on format means that the visual appearance of the normalized identifiers will not be changed as a result.

**Comment by jieyouxu** [comment]
NB. we now also NFC-normalize the `NAME` part of `'NAME` lifetimes since after https://github.com/rust-lang/rust/pull/149192.

**Comment by jieyouxu** [comment]
Thought: one slight caveat *might* be if `rustc` bumps its `unicode-normalization` some time in the future while `rustfmt` stays on an older version. Not entirely sure if that's an actual problem though. (It *might* be slightly annoying, because say this option is stabilized, then we may have to do significant `unicode-normalization` bumps only over a style edition gate... (?))

**Comment by Jules-Bertholet** [comment]
Programs containing non-ASCII characters are currently exempt from rustfmt's stability guarantees, as stated in the README.
